### ISSUE TITLE
Policy migrations fail from v10.13.2 to v11.0.0 with MySQL 5.7

### ISSUE THREAD

**Issue by arnediekmann** [directus/directus#23234]
### Describe the Bug

When updating a Directus cluster from version `10.13.2` to `11.0.0` the migrations fail with the following error:

```
[11:50:58.185] INFO: Applying Permissions Policies...
Error: alter table `directus_permissions` modify  `policy` char(36) not null  - Cannot change column 'policy': used in a foreign key constraint 'directus_permissions_policy_foreign'
    at Packet.asError (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/packets/packet.js:728:17)
    at Query.execute (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/commands/command.js:29:26)
    at Connection.handlePacket (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/connection.js:481:34)
    at PacketParser.onPacket (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/connection.js:97:12)
    at PacketParser.executeStart (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/packet_parser.js:75:16)
    at Socket.<anonymous> (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/connection.js:104:25)
    at Socket.emit (node:events:517:28)
    at addChunk (node:internal/streams/readable:368:12)
    at readableAddChunk (node:internal/streams/readable:341:9)
    at Readable.push (node:internal/streams/readable:278:10) {
  code: 'ER_FK_COLUMN_CANNOT_CHANGE',
  errno: 1832,
  sqlState: 'HY000',
  sqlMessage: "Cannot change column 'policy': used in a foreign key constraint 'directus_permissions_policy_foreign'",
  sql: 'alter table `directus_permissions` modify  `policy` char(36) not null '
}
```

Afterwards the API is only responding with error 500. In the logs I see the following error:

```
[13:32:04.892] ERROR: select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is nul
l - Table 'xx_dev_directus.directus_access' doesn't exist
    err: {
      "type": "Error",
      "message": "select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null -
Table 'xx_dev_directus.directus_access' doesn't exist",
      "stack":
          Error: select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null - T
able 'xx_dev_directus.directus_access' doesn't exist
              at Packet.asError (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/packets/packet.js:728:17)
              at Query.execute (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/commands/command.js:29:26)
              at Connection.handlePacket (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/connection.js:481:34)
              at PacketParser.onPacket (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/connection.js:97:12)
              at PacketParser.executeStart (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/packet_parser.js:75:16)
              at Socket.<anonymous> (/directus/node_modules/.pnpm/mysql2@3.10.0/node_modules/mysql2/lib/connection.js:104:25)
              at Socket.emit (node:events:517:28)
              at addChunk (node:internal/streams/readable:368:12)
              at readableAddChunk (node:internal/streams/readable:341:9)
              at Readable.push (node:internal/streams/readable:278:10)
      "code": "ER_NO_SUCH_TABLE",
      "errno": 1146,
      "sqlState": "42S02",
      "sqlMessage": "Table 'xx_dev_directus.directus_access' doesn't exist",
      "sql": "select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null"
    }
```

When inspecting the tables that [the migration](https://github.com/directus/directus/blob/main/api/src/database/migrations/20240806A-permissions-policies.ts) changes (`directus_access`, `directus_permissions`, `directus_policies` and `directus_users`) I see the following state:

```
mysql> describe directus_access;
ERROR 1146 (42S02): Table 'xx_dev_directus.directus_access' doesn't exist

mysql> describe directus_policies;
+--------------+--------------+------+-----+---------+-------+
| Field        | Type         | Null | Key | Default | Extra |
+--------------+--------------+------+-----+---------+-------+
| id           | char(36)     | NO   | PRI | NULL    |       |
| name         | varchar(100) | NO   |     | NULL    |       |
| icon         | varchar(64)  | NO   |     | badge   |       |
| description  | text         | YES  |     | NULL    |       |
| ip_access    | text         | YES  |     | NULL    |       |
| enforce_tfa  | tinyint(1)   | NO   |     | 0       |       |
| admin_access | tinyint(1)   | NO   |     | 0       |       |
| app_access   | tinyint(1)   | NO   |     | 0       |       |
+--------------+--------------+------+-----+---------+-------+
8 rows in set (0.00 sec)

mysql> describe directus_permissions;
+-------------+------------------+------+-----+---------+----------------+
| Field       | Type             | Null | Key | Default | Extra          |
+-------------+------------------+------+-----+---------+----------------+
| id          | int(10) unsigned | NO   | PRI | NULL    | auto_increment |
| collection  | varchar(64)      | NO   | MUL | NULL    |                |
| action      | varchar(10)      | NO   |     | NULL    |                |
| permissions | json             | YES  |     | NULL    |                |
| validation  | json             | YES  |     | NULL    |                |
| presets     | json             | YES  |     | NULL    |                |
| fields      | text             | YES  |     | NULL    |                |
| policy      | char(36)         | YES  | MUL | NULL    |                |
| role        | char(36)         | YES  |     | NULL    |                |
+-------------+------------------+------+-----+---------+----------------+
9 rows in set (0.00 sec)

mysql> describe directus_users;
+-----------------------+--------------+------+-----+-------------------+-----------------------------+
| Field                 | Type         | Null | Key | Default           | Extra                       |
+-----------------------+--------------+------+-----+-------------------+-----------------------------+
| id                    | char(36)     | NO   | PRI | NULL              |                             |
| first_name            | varchar(50)  | YES  |     | NULL              |                             |
| last_name             | varchar(50)  | YES  |     | NULL              |                             |
| email                 | varchar(128) | YES  | UNI | NULL              |                             |
| password              | varchar(255) | YES  |     | NULL              |                             |
| location              | varchar(255) | YES  |     | NULL              |                             |
| title                 | varchar(50)  | YES  |     | NULL              |                             |
| description           | text         | YES  |     | NULL              |                             |
| tags                  | json         | YES  |     | NULL              |                             |
| avatar                | char(36)     | YES  |     | NULL              |                             |
| language              | varchar(255) | YES  |     | NULL              |                             |
| tfa_secret            | varchar(255) | YES  |     | NULL              |                             |
| status                | varchar(16)  | NO   |     | active            |                             |
| role                  | char(36)     | YES  | MUL | NULL              |                             |
| token                 | varchar(255) | YES  | UNI | NULL              |                             |
| last_access           | timestamp    | NO   |     | CURRENT_TIMESTAMP | on update CURRENT_TIMESTAMP |
| last_page             | varchar(255) | YES  |     | NULL              |                             |
| provider              | varchar(128) | NO   |     | default           |                             |
| external_identifier   | varchar(255) | YES  | UNI | NULL              |                             |
| auth_data             | json         | YES  |     | NULL              |                             |
| email_notifications   | tinyint(1)   | YES  |     | 1                 |                             |
| appearance            | varchar(255) | YES  |     | NULL              |                             |
| theme_dark            | varchar(255) | YES  |     | NULL              |                             |
| theme_light           | varchar(255) | YES  |     | NULL              |                             |
| theme_light_overrides | json         | YES  |     | NULL              |                             |
| theme_dark_overrides  | json         | YES  |     | NULL              |                             |
+-----------------------+--------------+------+-----+-------------------+-----------------------------+
26 rows in set (0.00 sec)
```

MySQL Server info:

```
+-------------------------+-------------------------------------------------------------------------------------------------+
| Variable_name           | Value                                                                                           |
+-------------------------+-------------------------------------------------------------------------------------------------+
| innodb_version          | 5.7.41-44                                                                                       |
| protocol_version        | 10                                                                                              |
| slave_type_conversions  |                                                                                                 |
| tls_version             | TLSv1,TLSv1.1,TLSv1.2                                                                           |
| version                 | 5.7.41-44-57-log                                                                                |
| version_comment         | Percona XtraDB Cluster (GPL), Release rel44, Revision b46a1c3, WSREP version 31.65, wsrep_31.65 |
| version_compile_machine | x86_64                                                                                          |
| version_compile_os      | debian-linux-gnu                                                                                |
| version_suffix          | -57-log                                                                                         |
+-------------------------+-------------------------------------------------------------------------------------------------+
```

### To Reproduce

1. Setup Directus instance in version `10.13.2`
2. Update Directus image tag to `11.0.0`

### Directus Version

v11.0.0

### Hosting Strategy

Self-Hosted (Docker Image)

### Database

Percona XtraDB Cluster 5.7.41

**Comment by hanneskuettner** [comment]
> When inspecting the tables that [the migration](https://github.com/directus/directus/blob/main/api/src/database/migrations/20240806A-permissions-policies.ts) changes (directus_access, directus_permissions, directus_policies and directus_users) I see the following state:

That is after the migration failed mid-way through?

**Comment by hanneskuettner** [comment]
> `| version                 | 5.7.41-44-57-log   `

Also that looks suspiciously like a MySQL v5.7 which isn't officially supported by us anymore, as it reached EOL a while ago now.

**Comment by arnediekmann** [comment]
Thanks for the swift reply. 

Yes, it is indeed a MySQL 5.7 server which is due for update sometime soon - perhaps we can speed that update up and thereby make it work. 

But that would also explain why it worked locally: that database server is running on MySQL 8. 

If there was a workaround to make 5.7 work in the meantime, we would be eager to try it, though 😉 

I already tried to execute the migrations manually but since that seemed tedious, I instead submitted the issue here 😇

**Comment by arnediekmann** [comment]
> That is after the migration failed mid-way through?

Exactly - seemingly some of the operations succeeded, but not all.

**Comment by hanneskuettner** [comment]
> Exactly - seemingly some of the operations succeeded, but not all.

The line that fails is this one 

https://github.com/directus/directus/blob/15eebdb87585cb4af33e8e3b1754acb0ed555532/api/src/database/migrations/20240806A-permissions-policies.ts#L249

I'd be open to drop the foreign constraint before removing the nullable and then re-adding it after. 

I'll do some local testing with MySQL 5.7 to see if it's a 5.7 problem only.

**Comment by arnediekmann** [comment]
Thanks a ton - if there is a way for me to test this or if I should try creating a PR  with this, then let me know!

**Comment by fchavant-cl** [comment]
For reference: the same error occurs with Percona Server for MySQL 8.0 (8.0.37).

**Comment by arnediekmann** [comment]
Thanks for the work on #23327! 🥳

**Comment by cliqer** [comment]
Please reopen. This is still a problem with `MariaDB v10.11.9` upgrading from `Directus v10.13.2` to `v11.1.0`.
```
[00:49:29.986] ERROR: select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null - Table 'directusapi.directus_access' doesn't exist
```
I have migrated down back to 10.13.2 but something is messed up with my migration and `directus_policies` is not being created again.

**Comment by kevincam3** [comment]
I'm still not able to upgrade either. It always breaks and doesn't create the directus_access table

**Comment by hanneskuettner** [comment]
> I'm still not able to upgrade either. It always breaks and doesn't create the directus_access table

Can you give more detail as to what breaks? It should fail in the migration when trying to migrate a v10 database to v11. This error should be reported in the console. With this error we might be able to deduct what is breaking in your case, since, even though the result is the same (`directus_access` is missing) it is highly unlikely it breaks in the same way as the original issue, as it has been fixed.

**Comment by cliqer** [comment]
Still this is an issue with v11.2.0.

After upgrade and migration from v10.13.2, this is the log.
Errors as reported before and no new tables are being generated:
```
[07:23:27.397] WARN: Some tables and columns do not match your database's default collation (utf8mb3_general_ci):
[07:23:31.688] ERROR: select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null - Table 'api_test.directus_access' doesn't exist
    err: {
      "type": "Error",
      "message": "select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null - Table 'api_test.directus_access' doesn't exist",
      "stack":
          Error: select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null - Table 'api_test.directus_access' doesn't exist
              at Packet.asError (/var/www/vhosts/example.com/api.example.com/node_modules/mysql2/lib/packets/packet.js:738:17)
              at Query.execute (/var/www/vhosts/example.com/api.example.com/node_modules/mysql2/lib/commands/command.js:29:26)
              at Connection.handlePacket (/var/www/vhosts/example.com/api.example.com/node_modules/mysql2/lib/connection.js:481:34)
              at PacketParser.onPacket (/var/www/vhosts/example.com/api.example.com/node_modules/mysql2/lib/connection.js:97:12)
              at PacketParser.executeStart (/var/www/vhosts/example.com/api.example.com/node_modules/mysql2/lib/packet_parser.js:75:16)
              at Socket.<anonymous> (/var/www/vhosts/example.com/api.example.com/node_modules/mysql2/lib/connection.js:104:25)
              at Socket.emit (node:events:519:28)
              at addChunk (node:internal/streams/readable:559:12)
              at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)
              at Readable.push (node:internal/streams/readable:390:5)
      "code": "ER_NO_SUCH_TABLE",
      "errno": 1146,
      "sqlState": "42S02",
      "sqlMessage": "Table 'api_test.directus_access' doesn't exist",
      "sql": "select `directus_policies`.`ip_access` as `ip_access` from `directus_access` left join `directus_policies` on `directus_access`.`policy` = `directus_policies`.`id` where `directus_policies`.`ip_access` is not null and `role` is null and `user` is null"
    }
```

**Comment by alexey-yarmosh** [comment]
Here is the log of the error that happens during the migration from v10 to v11 for MariaDB:

```
globalping-dash-directus-1  | [18:03:46.177] TRACE: [1.479ms] alter table `directus_permissions` drop foreign key `directus_permissions_role_foreign` []
globalping-dash-directus-1  | [18:03:46.178] TRACE: [0.444ms] update `directus_permissions` set `role` = ? where `role` is null [abf8a154-5b1c-4a46-ac9c-7300570f4f17]
globalping-dash-directus-1  | [18:03:46.179] TRACE: [0.805ms] update `directus_permissions` set `policy` = `role` []
globalping-dash-directus-1  | [18:03:46.182] TRACE: [2.763ms] alter table `directus_permissions` drop `role` []
globalping-dash-directus-1  | [18:03:46.184] TRACE: [1.191ms] select * from information_schema.columns where table_name = ? and table_schema = ? [directus_permissions, dashboard-globalping]
globalping-dash-directus-1  | [18:03:46.188] TRACE: [3.808ms] alter table `directus_permissions` modify  `policy` char(36) not null  []
globalping-dash-directus-1  | [18:03:46.188] TRACE: [5.799ms] SELECT 1 []
globalping-dash-directus-1  | Error: alter table `directus_permissions` add constraint `directus_permissions_policy_foreign` foreign key (`policy`) references `directus_policies` (`id`) on delete CASCADE - Can't create table `dashboard-globalping`.`directus_permissions` (errno: 150 "Foreign key constraint is incorrectly formed")
globalping-dash-directus-1  |     at Packet.asError (/directus/node_modules/.pnpm/mysql2@3.11.2/node_modules/mysql2/lib/packets/packet.js:738:17)
globalping-dash-directus-1  |     at Query.execute (/directus/node_modules/.pnpm/mysql2@3.11.2/node_modules/mysql2/lib/commands/command.js:29:26)
globalping-dash-directus-1  |     at Connection.handlePacket (/directus/node_modules/.pnpm/mysql2@3.11.2/node_modules/mysql2/lib/connection.js:481:34)
globalping-dash-directus-1  |     at PacketParser.onPacket (/directus/node_modules/.pnpm/mysql2@3.11.2/node_modules/mysql2/lib/connection.js:97:12)
globalping-dash-directus-1  |     at PacketParser.executeStart (/directus/node_modules/.pnpm/mysql2@3.11.2/node_modules/mysql2/lib/packet_parser.js:75:16)
globalping-dash-directus-1  |     at Socket.<anonymous> (/directus/node_modules/.pnpm/mysql2@3.11.2/node_modules/mysql2/lib/connection.js:104:25)
globalping-dash-directus-1  |     at Socket.emit (node:events:517:28)
globalping-dash-directus-1  |     at addChunk (node:internal/streams/readable:368:12)
globalping-dash-directus-1  |     at readableAddChunk (node:internal/streams/readable:341:9)
globalping-dash-directus-1  |     at Readable.push (node:internal/streams/readable:278:10) {
globalping-dash-directus-1  |   code: 'ER_CANT_CREATE_TABLE',
globalping-dash-directus-1  |   errno: 1005,
globalping-dash-directus-1  |   sqlState: 'HY000',
globalping-dash-directus-1  |   sqlMessage: 'Can\'t create table `dashboard-globalping`.`directus_permissions` (errno: 150 "Foreign key constraint is incorrectly formed")',
globalping-dash-directus-1  |   sql: 'alter table `directus_permissions` add constraint `directus_permissions_policy_foreign` foreign key (`policy`) references `directus_policies` (`id`) on delete CASCADE'
globalping-dash-directus-1  | }
mariadb-1                   | 2024-11-13 18:03:46 5 [Warning] Aborted connection 5 to db: 'dashboard-globalping' user: 'directus' host: '192.168.65.1' (Got an error reading communication packets)
mariadb-1                   | 2024-11-13 18:03:46 4 [Warning] Aborted connection 4 to db: 'dashboard-globalping' user: 'directus' host: '192.168.65.1' (Got an error reading communication packets)
globalping-dash-directus-1 exited with code 1
```

**Comment by hanneskuettner** [comment]
Or could this be related to this bug report in MariaDB? https://jira.mariadb.org/browse/MDEV-29717?
Can you check what the index type for the primary key on `directus_permissions` is on your DB?

**Comment by alexey-yarmosh** [comment]
Indexes:
<img width="860" alt="image" src="https://github.com/user-attachments/assets/5843aceb-d132-4b3c-a2d5-fcf5c49d104d">
<img width="848" alt="image" src="https://github.com/user-attachments/assets/199071ad-dc6c-4207-ae1e-bd18a7ac5190">

**Comment by cliqer** [comment]
On my case: `MariaDB v10.11.10`

State after failed migration to latest and migrate:down. (Now migrations do not seem to change anything if I retry to upgrade).

directus_permissions:
```sql
CREATE TABLE `directus_permissions` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `role` char(36) DEFAULT NULL,
  `collection` varchar(64) NOT NULL,
  `action` varchar(10) NOT NULL,
  `permissions` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `validation` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `presets` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `fields` text DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `directus_permissions_collection_foreign` (`collection`),
  KEY `directus_permissions_role_foreign` (`role`),
  CONSTRAINT `directus_permissions_role_foreign` FOREIGN KEY (`role`) REFERENCES `directus_roles` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=694 DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_general_ci;
```
directus_roles:
```sql
CREATE TABLE `directus_roles` (
  `id` char(36) NOT NULL,
  `name` varchar(100) NOT NULL,
  `icon` varchar(64) NOT NULL DEFAULT 'supervised_user_circle',
  `description` text DEFAULT NULL,
  `ip_access` text DEFAULT NULL,
  `enforce_tfa` tinyint(1) NOT NULL DEFAULT 0,
  `admin_access` tinyint(1) NOT NULL DEFAULT 0,
  `app_access` tinyint(1) NOT NULL DEFAULT 1,
  `parent` char(36) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_general_ci;
```
No directus_policies is being created on upgrade.

Please let me know if you need more info.

Also, could someone please share a directus policies schema for mariadb that works as to try to create manually and test?

**Comment by alexey-yarmosh** [comment]
I found the problem. Our DB default charset is `utf8mb4` and collation is **`utf8mb4_unicode_ci`**. It was working fine, but some time ago we had to add `DB_CHARSET=utf8mb4` env var to support emojis in notifications. It turned out that when `DB_CHARSET=utf8mb4` is set directus will create new tables with the **`utf8mb4_general_ci`** collation.

Now it is the situation where `directus_permissions.policy` has collation `utf8mb4_unicode_ci` as that table was created previously, and new `directus_policies.id` has collation `utf8mb4_general_ci`. Which leads to error during adding foreign key constraint.

I didn't find a way to tell directus to use `utf8mb4_unicode_ci`. So not sure how the fix should look like.
